Tweak JSON decoding (to reduce its footprint and make more compositional)

matthew-hilty opened the pull request:
[TL;DR: I've tried making some of the JSON decoding sections more succinct. Let me know if you think any of my suggestions are helpful. Thanks.]

I did a pull request a while back, where I suggested some changes for profile decoding because the node-express backend played a little loose with the RealWorld/Conduit spec. My changes, however, weren't very "elegant". Maybe they were helpful, but abandoning Argonaut's automatic JSON decoding, to the extent it had been available, and replacing it with overly specific decoders definitely had its tradeoffs: it added clutter and it made the overall codebase less compositional and seamless.

I've been working on some decoding tools since then, just simple utility functions to help with Argonaut, and I think they help with some of this frontend's decoding use cases -- in particular, handling JSON data with optional field absence. In case you might be interested, I've wired up some of these utilities in your application for you to check out.

-----
##### Some info about the changes:

The library 'purescript-tolerant-argonaut' is essentially a reworking of standard Argonaut, except it treats simple `Plus` instances differently, interpreting any JSON field with an absent `Maybe` or `Array` value as that instance's `empty` value.

The library offers a couple of other things that also seem to expedite or "clean up" common decoding patterns. For example, I've noticed that this Conduit frontend often has to manually decode JSON input into all the fields of a datatype because of one or two complexish, wayward fields for which default Argonaut decoding doesn't work. 'Tolerant-argonaut' makes two utilities available (`decodeJsonWith` and `decodeJsonPer`) that skip past all the well-behaved fields, delegating to Argonaut to handle them, and that allow customized decoders for just the unruly fields.

An example of `decodeJsonWith` is used in 'Profile.purs'. In addition to

```purescript
type Author' = { username :: Username, bio :: Maybe String, image :: Maybe Avatar, following :: Relation }

decodeAuthor' :: Maybe Username -> Json -> Either String Author'
decodeAuthor' mbUsername json = do 
  obj <- decodeJson json
  username <- obj .: "username"
  bio <- obj .: "bio"
  image <- obj .: "image"
  following <- obj .: "following"
  let profile = { username, bio, image, following }
  pure $ case mbUsername of
    Just currentUsername | username == currentUsername -> profile { following = You }
    _ -> profile
```

... the following alternative is also possible.

```purescript
decodeAuthor' :: Maybe Username -> Json -> Either String Author'
decodeAuthor' mbUsername = decodeJsonWith { following: getFollowing }
  where
  getFollowing followingJson authorRecord = case mbUsername of
    Just currentUsername | authorRecord.username == currentUsername -> pure You
    _ -> decodeJson followingJson
```

The actual code used is a bit different because of `Author`'s fieldname change, but the preceding codeblock gives the gist.

matthew-hilty committed:
Use 'tolerant-argonaut' package to facilitate decoding.

thomashoneyman commented on src/Data/Utils.purs:
````diff
@@ -0,0 +1,31 @@
+-- | ```
+
+decodeAt :: forall a. Tol.DecodeJson a => String -> Json -> Either String a
+decodeAt key json = ((_ .:: key) <=< decodeJson) json
````
What about:

```suggestion
decodeAt key = (_ .:: key) <=< decodeJson
```

thomashoneyman commented on src/Data/Utils.purs:
```diff
@@ -0,0 +1,31 @@
+import Data.Argonaut.Core (Json)
+import Data.Argonaut.Decode (decodeJson)
+import Data.Argonaut.Decode.Struct.Tolerant ((.::))
+import Data.Argonaut.Decode.Struct.Tolerant (class DecodeJson) as Tol
```
Could we standardize on importing non-operators from `Data.Decode.Struct.Tolerant` as:

```suggestion
import Data.Argonaut.Decode.Struct.Tolerant as Tolerant
```

...just for uniformity's sake?

thomashoneyman commented on src/Data/Article.purs:
```diff
@@ -74,33 +74,58 @@ type ArticleWithMetadata = { | ArticleRep + ArticleMetadataRep () }
+-- |   where
+-- |   renameFields = rrenameMany { articles: SProxy :: SProxy "body"
+-- |                              , articlesCount: SProxy :: SProxy "total"
+-- |                              }
```
Minor style nit: mind shifting this to newlines to conserve horizontal space?

```purescript
  renameFields =
    rrenameMany
      { articles: SProxy :: SProxy "body"
      , articlesCount: SProxy :: SProxy "total"
      }
```

thomashoneyman commented on bower.json:
```diff
@@ -26,12 +26,23 @@
+    "purescript-argonaut-core": "^5.0.0",
+    "purescript-profunctor-lenses": "^6.2.0",
+    "purescript-record": "^2.0.0",
+    "purescript-foreign-object": "^2.0.0"
```
This file can be removed entirely now that the project is using Spago (see #26)

thomashoneyman commented on src/Data/Profile.purs:
```diff
@@ -20,14 +20,20 @@ import Prelude
 import Data.Argonaut.Core (Json)
-import Data.Argonaut.Decode (class DecodeJson, decodeJson, (.:), (.:?))
+import Data.Argonaut.Decode (class DecodeJson, decodeJson)
+import Data.Argonaut.Decode.Struct.Tolerant (decodeJsonPer, decodeJsonWith) as Tolerantly
```
Same here with importing just as `Tolerant`:

```suggestion
import Data.Argonaut.Decode.Struct.Tolerant as Tolerant
```

thomashoneyman reviewed (changes requested):
Thank you for making this improvement, and for waiting patiently for me to get back to you. Sorry for the delay! I've got a few comments I'd like addressed, and this will have to be (lightly) updated with the master branch, but then I'm happy to merge.

matthew-hilty commented:
Thanks for being open to using my suggestions. And sorry for my own delayed response -- I've been meaning to make the code changes you've requested (as well as some other code fixes to accommodate the new 'typelevel-prelude'), but it looks like Pursuit hasn't been updated in a while, so  it refuses to publish packages when it encounters kind annotations in 'forall' contexts.

Right now, I'm thinking I'll wait about a week to see if Pursuit gets rebuilt, but if not, I'll unwind code for backward compatibility. Either way, though, I should be able to push your requested changes soon.

thomashoneyman commented:
@matthew-hilty Thanks! Feel free to make the changes here even without corresponding updates to Pursuit -- it doesn't actually host packages, so dependencies will still work fine despite not being uploaded to Pursuit.

matthew-hilty committed:
Merge branch 'master' into tweak-decoding

matthew-hilty committed:
Continue merging branch 'master' into tweak-decoding
by updating the *.dhall config files and the typelevel-prelude module imports

matthew-hilty committed:
Make requested changes:
standardize module qualifier for tolerant-argonaut imports
and eta-reduce 'decodeAt'.

matthew-hilty committed:
Refactor Profile.purs: remove unhelpful decodeArticle' function.

matthew-hilty commented:
I've made the changes you've requested. Let me know if you'd like any other revisions.

Also, thanks for the Pursuit tip -- spago sure makes package management quick and easy.

matthew-hilty committed:
Update spago bundling command to 'bundle-app'.

thomashoneyman commented:
@matthew-hilty Thanks for contributing this update, and I appreciate your work on it and the libraries underneath. I'm a little worried about introducing the extra dependencies, especially since they're a mix of fairly new pre-1.0 and 1.0 libraries. 

Would you mind keeping an eye on this project when you make updates to those libraries -- and removing the package overrides if your libraries are added to the main PureScript package set? That would help me feel a little more comfortable bringing in a few libraries I'm unfamiliar with.

thomashoneyman commented on packages.dhall:
```diff
@@ -57,6 +57,70 @@ let additions =
           ]
           "https://github.com/awakesecurity/purescript-precise-datetime.git"
           "v5.1.1"
+      , higher-order =
```
You may be able to add these to https://github.com/purescript/package-sets so that they don't need to be maintained as explicit overrides in the future. Not a problem for this PR, but that makes things a little more convenient for folks downstream, as they can then just do `spago install <lib>`.

matthew-hilty commented:
> I'm a little worried about introducing the extra dependencies, especially since they're a mix of fairly new pre-1.0 and 1.0 libraries.

That's understandable. I see that you've already accepted the PR, but if you'd prefer to roll back and wait until the projects are more mature, I wouldn't take offense. Alternatively, if any changes to 'tolerant-argonaut' (TA) or any of its dependencies might give you greater confidence, I'd be inclined to make them. My goal has been to author useful infrastructural libraries, tailored to meet the needs of applications like "Real World Halogen", so, of course, any feedback by those applications' creators would be welcome (if they want and if time permits).

> Would you mind keeping an eye on this project when you make updates to those libraries -- and removing the package overrides if your libraries are added to the main PureScript package set?

Absolutely. This project is actually what inspired development of TA, and I've been planning to add TA and its dependencies to spago's package sets.

thomashoneyman commented:
I'm glad to hear all of that. I'm going to go ahead and merge this, and I look forward to exploring these libraries more. I'll be sure to give you feedback if I notice anything in particular. Thanks again for your work on this!
